common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitUsbSmartcardCertBasedAuthManager.java

```diff
@@ -0,0 +1,204 @@
+            mUsbYubiKeyManager.disable();
+        }
+    }
+
+    /**
+     * Request a PivSession instance in order to carry out methods
+     *  implemented in YubiKitSmartcardSession.
+     * @param callback Contains callbacks to run when a PivSession is successfully instantiated and when any exception is thrown due to a connection issue.
+     */
+    @Override
+    void requestDeviceSession(@NonNull ISessionCallback callback) {
+        final String methodTag = TAG + "requestDeviceSession:";
+        synchronized (sDeviceLock) {
+            if (isDeviceConnected()) {
+                //Request a connection from mUsbDevice so that we can get a PivSession instance.
+                mUsbDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+                    @Override
+                    public void invoke(@NonNull final Result<UsbSmartCardConnection, IOException> value) {
+                        try {
+                            final SmartCardConnection c = value.getValue();
+                            final PivSession piv = new PivSession(c);
+                            final YubiKitSmartcardSession session = new YubiKitSmartcardSession(piv);
+                            callback.onGetSession(session);
+                        } catch (final Exception e) {
+                            callback.onException(e);
+                        }
+                    }
+                });
+                return;
+            }
+            Logger.error(methodTag, DEVICE_ERROR_MESSAGE, null);
+            callback.onException(new Exception());
+        }
+    }
+
+    /**
+     * Checks if a YubiKey is currently connected via USB.
+     * @return true if YubiKey is currently connected. Otherwise, false.
+     */
+    @Override
+    boolean isDeviceConnected() {
+        synchronized (sDeviceLock) {
+            return mUsbDevice != null;
+        }
+    }
+
+    /**
+     * Runs implementation specific processes that may need to occur just before calling {@link ClientCertRequest#proceed(PrivateKey, X509Certificate[])}.
+     * @param telemetryHelper CertBasedAuthTelemetryHelper instance.
+     */
+    @Override
+    void initBeforeProceedingWithRequest(@NonNull CertBasedAuthTelemetryHelper telemetryHelper) {
+        YubiKeyPivProviderManager.addPivProvider(telemetryHelper, getPivProviderCallback());
+    }
+
+    /**
+     * Cleanup to be done upon host activity being destroyed.
+     * @param activity current host activity.
+     */
+    @Override
+    void onDestroy(@NonNull final Activity activity) {
+        stopDiscovery(activity);
+    }
+
+    /**
+     * Used to provide PivProvider constructor a Callback that will establish a new PivSession when it is needed.
+     * @return A Callback which returns a Callback that will return a new PivSession instance.
+     */
+    @NonNull
+    Callback<Callback<Result<PivSession, Exception>>> getPivProviderCallback() {
```

Comment:
maybe this could be moved to the base class?

```Java
protected Callback<Callback<Result<PivSession, Exception>>> getPivProviderCallback(YubiKeyDevice device) {
...
     yubikeyDevice.requestConnection()
{
```

Reply:
The base classes are now based on usb vs nfc instead of by smartcard brand, so the abstract class for YubiKit in general doesn't exist anymore. This change was done after logic needed to be added for determining if a previousy connected device was the same as the current, only for nfc and not usb.
I could maybe move this to the 'YubiKeyPivProviderManager' class though, and change the parameters of `requestConnection` based on an `instanceof` check (they change for nfc vs usb)

Reply:
That will basically just create a method that combines both `getPivProviderCalback` methods though, with an `if` statement. Esp since I would have to pass `YubiKeyDevice` as a parameter, not sure if it's an improvement?